lib/full-site-editing/block-templates.php

```diff
@@ -93,18 +113,62 @@ function _gutenberg_get_template_files( $template_type ) {
 				// Subtract ending '.html'.
 				-5
 			);
-			$template_files[] = array(
+
+			$new_template_item = array(
 				'slug'  => $template_slug,
 				'path'  => $template_file,
 				'theme' => $theme_slug,
 				'type'  => $template_type,
 			);
+
+			if ( 'wp_template_part' === $template_type ) {
+				$template_files[] = _gutenberg_conditionally_add_template_part_section_info(
+					$new_template_item,
+					$template_part_data
+				);
+			} else {
+				$template_files[] = $new_template_item;
+			}
 		}
 	}
 
 	return $template_files;
 }
 
+/**
+ * Attempts to read the theme.json and return the 'template-parts' field.
+ *
+ * @return array Template part data from theme.json or empty array if not found.
+ */
+function _gutenberg_get_template_part_info_from_theme_json() {
+	if ( is_readable( locate_template( 'experimental-theme.json' ) ) ) {
```

Comment:
I think the advantage of `locate_template` is that is accounts for parent themes and falls back to TEMPLATEPATH after STYLESHEETPATH.  This is similar to how global styles checks the theme JSON:

https://github.com/WordPress/gutenberg/blob/537412fd439c0c8a9befcf99e8bbbc2fdc8c1147/lib/global-styles.php#L13-L15

Now looking back, it seems like we should probably be using that global styles function here?

I don't think our problem with STYLESHEETPATH as noted in that issue exists here, since this gets called much later in the process than that example. 🤔

Reply:
I think the behavior should be the same in both cases. The benefits of not using `locate_template` would be:

- not calling the old code that's supposed to handle legacy themes (not applicable to FSE themes);
- `get_stylesheet_directory` is filterable. 

> Now looking back, it seems like we should probably be using that global styles function here?

Agreed. I think it would be better to reuse the existing global styles function. I'm still thinking that it could be refactored to use `get_stylesheet_directory` too (cc @nosolosw)

Reply:
:wave: I'd like to share a different perspective on data should be read out of theme.json. This is how I envision it working:

- Every new field is added to [the schema](https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-theme-json.php#L105) we have.
- The class `WP_Theme_JSON` is used to get data out of it. If necessary, by creating new getter methods. Example: `get_section` (just an example, not sure if it's a good name yet).

This approach has many benefits, including that we only read the theme.json once (I/O operation, costly) and that the interaction with the underlying data is abstracted away (keys can be moved to a different place without having to touch multiple files, the hard parts like merging theme/user data is done automatically for all fields that need it, etc).

As a follow-up to the recent refactors https://github.com/WordPress/gutenberg/pull/28110 https://github.com/WordPress/gutenberg/pull/28533 I had in my queue create a PR to migrate the existing `customTemplates` to work like that. Happy to accelerate that PR if it serves as inspiration for this one.

Reply:
>The class WP_Theme_JSON is used to get data out of theme.json. If necessary, by creating new getter methods. Example: get_section (just an example, not sure if it's a good name yet).

I think that makes a lot of sense!  Will see if I can get something like that working.  So we have an instance of `WP_Theme_JSON` already set up and just use a get method on it to get our information (instead of us having to decode in the request here)?

>Happy to accelerate that PR if it serves as inspiration for this one.

Im always open for more inspiration! 😆  - But no pressure if you have other things going on.

Reply:
ok, took a quick stab at this and prepared https://github.com/WordPress/gutenberg/pull/28700 

I still need to understand how `pageTemplates` is being used and whether other things need to be added, but it looks promising.

Reply:
@nosolosw - Thanks!  So something along the lines of this: 0659325 ?  (although eventually using the `get_theme_data` you are proposing in that PR as opposed to just using `get_theme_origin` and making it public like im doing here).

Reply:
Yeah. Would you help me review / land https://github.com/WordPress/gutenberg/pull/28700 so you can have it for this as well? I can polish it this morning.

Reply:
https://github.com/WordPress/gutenberg/pull/28700 is ready to land!

Reply:
So we did update this to work the JSON resolver which is great, but I think we might be getting a little distracted from @vindl 's original critique.  I think that was that the use of `locate_template` (and underlying `STYLESHEETPATH`) used in the json_resolver class, as opposed to using things like `get_stylesheet_path` that are filterable.  (Please free to clarify @vindl ).. But do you have thoughts regarding that @nosolosw ?

Reply:
ok, loaded my brain with these conversations. I see this comes from https://github.com/WordPress/gutenberg/pull/26500#discussion_r515047360 which ended up being problematic, hence why we did https://github.com/WordPress/gutenberg/pull/27021 What I've gathered is that the use of non-filterable constants (`STYLESHEET` instead of `get_stylesheet_directory`) is problematic under certain special setups, hence why `locate_template` is best to be avoided.

The use of `locate_template` in global styles boils down to support child themes that want to reuse the parent's `theme.json`. So, the change involves replicating part of the logic of `locate_template` but with methods instead of constants. Probably not necessary to have the theme-compat part. I can do that after https://github.com/WordPress/gutenberg/pull/28855 lands ― it needs a review 🙂

Reply:
Here's a bug report https://github.com/WordPress/gutenberg/issues/26359 that confirms that `locate_template` is problematic in certain contexts (customizer).

Reply:
https://github.com/WordPress/gutenberg/pull/28942 replaces `locate_template`.

Reply:
just need a :+1: to land https://github.com/WordPress/gutenberg/pull/28942 :)